mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulPlayScreen.kt

```diff
@@ -1,34 +1,141 @@
 package ch.epfl.sdp.mobile.state
 
 import androidx.compose.foundation.layout.PaddingValues
-import androidx.compose.runtime.Composable
-import androidx.compose.runtime.remember
+import androidx.compose.runtime.*
 import androidx.compose.ui.Modifier
+import ch.epfl.sdp.mobile.application.authentication.AuthenticatedUser
+import ch.epfl.sdp.mobile.application.chess.ChessFacade
+import ch.epfl.sdp.mobile.application.chess.Match
+import ch.epfl.sdp.mobile.application.chess.notation.serialize
+import ch.epfl.sdp.mobile.state.Loadable.Companion.loaded
+import ch.epfl.sdp.mobile.state.Loadable.Companion.loading
 import ch.epfl.sdp.mobile.ui.play.PlayScreen
 import ch.epfl.sdp.mobile.ui.play.PlayScreenState
+import ch.epfl.sdp.mobile.ui.social.ChessMatch
+import ch.epfl.sdp.mobile.ui.social.Tie
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
 
+/** An Loadable Union Type to differentiate between loaded and state */
+sealed interface Loadable<out T> {
+  object Loading : Loadable<Nothing>
+  data class Loaded<out T>(val value: T) : Loadable<T>
+
+  companion object {
+    fun loading(): Loadable<Nothing> = Loading
+    fun <T> loaded(value: T): Loadable<T> = Loaded(value)
+  }
+}
 /**
- * Player class that holds info about chess player
- * @param newGame Callable when new game button is actioned. Switches to pregame screen
+ * A map for the Loadable type
+ * @param f callback function for map
  */
-/* FIXME: add more useful args : user... */
-class Player(private val newGame: () -> Unit) : PlayScreenState {
-  /* TODO: override other attributes */
-  override val onNewGameClick = newGame
+inline fun <A, B> Loadable<A>.map(f: (A) -> B): Loadable<B> =
+    when (this) {
+      is Loadable.Loaded -> loaded(f(value))
+      Loadable.Loading -> loading()
+    }
```

Comment:
I don't think these should be exposed as `public`. These classes are just an implementation detail of the `PlayScreenStateImpl`.

Reply:
Need to be public in order to write tests for it

Reply:
Because it's an implementation detail, maybe the tests should not target the functions exposed in Loadable directly, but it should be tested indirectly when you check the behavior of the viewmodel.